server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -57,7 +88,15 @@ public final class RemoteStoreRefreshListener implements ReferenceManager.Refres
     private final RemoteSegmentStoreDirectory remoteDirectory;
     private final Map<String, String> localSegmentChecksumMap;
     private long primaryTerm;
-    private static final Logger logger = LogManager.getLogger(RemoteStoreRefreshListener.class);
+
+    /**
+     * Semaphore that ensures there is only 1 retry scheduled at any time.
+     */
+    private final Semaphore retrySemaphore = new Semaphore(MAX_CONCURRENT_SCHEDULED_REMOTE_REFRESH_RETRIES);
```

Comment:
Is it possible that normal refresh and retried refresh (of previous failed normal refresh) are happening at the same time?

Reply:
No, that wont happen as there is `syncSegments(boolean isRetry)` method  which is synchronised. But this can happen that a retry has been scheduled to happen after some x seconds, and there is a normal refresh that is happening now. If this normal refresh fails, it can not schedule another retry as the previous has already acquired the single permit.